Sprint 07

ArturKorobeynikov opened the pull request:
1. Добавлены Unit тесты (arrayTests, MovieLoaderTests, MovieQuizTests, MovieQuizPresenterTests)
2. Добавлены UI тесты (MovieQuizUITests, MovieQuizUITestsLauncTets)
3. Создана новая сущность MovieQuizPresenter (перенесена вся бизнес-логика приложения, переход от MVC -> MVP)

ArturKorobeynikov committed:
1. Выделены отдельные сущности AlertModel, QuizQuestion, QuizResultsViewModel, QuizStepViewModel
2. Выделен отдельно вызов алерта в AlertPresenter
3. Добавлена папка с сервисами, включая новый сервис с ведением статистики игры
4. Внесены изменения в MovieQuizViewController в части вынесения сущностей в отдельные файлы, а также добавление модификация алерта дополнительной статистикой

ArturKorobeynikov committed:
1. Убрал лишние комментарии
2. Добавил к выделенным сущностям приписку final
3. Удалил методы, которые не использовались
4. Поправил отображение скругления углов для UIImage, сделал, чтобы не слетели шрифты на кнопках

ArturKorobeynikov committed:
1. Добавлены новые сущности для загрузки данных из сети (файлы MostPopularMovies, MoviesLoader, NetworkClient)
2. Изменена структура QuizQuestion под потребности загрузки данных из сети
3. Добавлен индикатор загрузки в сторибоард и в MovieQuizViewController
4. Добавлены новые функции в MQVC (showLoadingIndicator, showNetworkError, didLoadDataFromServer, didFailToLoadData
5. Внесены изменения в сущности QuestionFactory (под нужды загрузки данных из сети)
6. Вопрос о рейтинге сделан псевдорандомным (с изменением числа рейтинга и вопроса "больше" или "меньше")

ArturKorobeynikov committed:
1. Добавлены Unit тесты (arrayTests, MovieLoaderTests, MovieQuizTests, MovieQuizPresenterTests)
2. Добавлены UI тесты (MovieQuizUITests, MovieQuizUITestsLauncTets)
3. Создана новая сущность MovieQuizPresenter (перенесена вся бизнес-логика приложения, переход от MVC -> MVP)

protos17 commented on MovieQuiz/AppDelegate.swift:
```diff
@@ -11,7 +11,6 @@ import UIKit
 final class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
```
Рекомендую лишние комментарии тоже поудалять

protos17 commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,22 @@
+//  Created by Артур Коробейников on 31.01.2023.
+//
+
+import Foundation
```
Foundation можно не испортировать, если импортируешь UIKit, потому что второй содержит в себе первый

protos17 commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,151 @@
+//  Created by Артур Коробейников on 09.03.2023.
+//
+
+import Foundation
```
Можно не импортировать Foundation

protos17 commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,151 @@
+            return
+        }
+        let givenAnswer = true
+        //viewController?.showAnswerResult(isCorrect: givenAnswer == currentQuestion.correctAnswer)
```
Неиспользуемый закомментированный код рекомендую удалить

protos17 commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,151 @@
+    }
+    
+    func proceedWithAnswer(isCorrect: Bool) {
+        //didAnswer(isYes: isCorrect)
```
Рекомендую удалить

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,194 +1,79 @@
+        super.viewDidLoad()
+        showLoadingIndicator()
+        presenter = MovieQuizPresenter(viewController: self)
+        //presenter.questionFactory?.loadData()
```
Рекомендую удалить

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -199,56 +84,56 @@ final class MovieQuizViewController: UIViewController {
+ 
  Картинка: Vivarium
  Настоящий рейтинг: 5,8
  Вопрос: Рейтинг этого фильма больше чем 6?
```
Рекомендую мок-данные совсем удалить из проекта

protos17 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,123 @@
+     image: "Vivarium",
+     text: "Рейтинг этого фильма больше чем 6?",
+     correctAnswer: false)]
+     */
```
Рекомендую удалить лишний закомментированный код

protos17 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,123 @@
+            do {
+                imageData = try Data(contentsOf: movie.resizedImageURL)
+            } catch {
+                print("Failed to load image")
```
В этом месте, функция не завершает свою работу и идет формировать вопрос без постера. 
В таком случае, бизнес-логика приложения будет нарушена, потому что пользователь не сможет ответить корректно, ведь изображения нет.
Это состояние стоит обработать, например, с помощью алерта. 
Не забудь, что алерт будет обновлять UI и это тоже нужно учесть (вернуть на main), ровно также, как и прерывание функции (сделать return)

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,96 @@
+        
+        var correctAnswers: Int {
+            get {
+                return (userDefaults.value(forKey: Keys.correct.rawValue) as? Int) ?? 0
```
Рекомендую здесь использовать userDefaults.integer(forKey: Keys.correct.rawValue)

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,96 @@
+        
+        var totalQuestions: Int {
+            get {
+                return (userDefaults.value(forKey: Keys.total.rawValue) as? Int) ?? 0
```
Рекомендую userDefaults.integer(forKey: Keys.correct.rawValue)

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,96 @@
+    var totalAccuracy: Double {
+        get {
+            let amountOfCorrect = (userDefaults.value(forKey: Keys.correct.rawValue) as? Double) ?? 0
+            let amountOfQuestions = (userDefaults.value(forKey: Keys.total.rawValue) as? Double) ?? 0
```
Рекомендую userDefaults.double(forKey: ...

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,96 @@
+    
+    var gamesCount: Int {
+        get {
+            let gamesCount = (userDefaults.value(forKey: Keys.gamesCount.rawValue) as? Int) ?? 0
```
Рекомендую userDefaults.integer(forKey: ...

protos17 commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,37 @@
+import XCTest
+@testable import MovieQuiz
+
+class ArrayTests: XCTestCase {
```
Рекомендую тоже проставлять final

protos17 commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,37 @@
+class ArrayTests: XCTestCase {
+    func testGetValueInRange() throws {
+        //Given
+        let array = [1, 1, 2, 3, 5]
```
рекомендую использовать уникальные значения, потому что array[safe: 0] и array[safe: 1] дадут один и тот же результат

protos17 commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    func testGetValueOutOfRange() throws {
+        //Given
+        let array = [1, 1, 2, 3, 5]
```
Рекомендую использовать уникальные значения

protos17 commented on MovieQuizTests/MovieLoaderTests.swift:
```diff
@@ -0,0 +1,105 @@
+    }
+}
+
+class MovieLoaderTests: XCTestCase {
```
Рекомендую проставить final

protos17 commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,102 @@
+
+import XCTest
+
+class MovieQuizUITests: XCTestCase {
```
Рекомендую проставить final

protos17 commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    func testYesButton() {
+        sleep(3)
```
Никак не влияет на тест, только задерживает его выполнение, можно удалить

protos17 commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+    
+    func testNoButton() {
+        sleep(3)
```
Никак не влияет на тест, только задерживает его выполнение, можно удалить

protos17 commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+    
+    func testGameFinish() {
+        sleep(2)
```
Никак не влияет на тест, только задерживает его выполнение, можно удалить

protos17 commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+    
+    func testAlertDismiss() {
+        sleep(2)
```
Никак не влияет на тест, только задерживает его выполнение, можно удалить

protos17 commented on MovieQuizUITests/MovieQuizUITestsLaunchTests.swift:
```diff
@@ -0,0 +1,32 @@
+
+import XCTest
+
+class MovieQuizUITestsLaunchTests: XCTestCase {
```
Рекомендую проставить final

ArturKorobeynikov committed:
1. удалены лишние комментарии
2. проставлены final к классам
3. удален import foundation, где был добавлен UIKit
